src/commands/filtercommands.cpp

```diff
@@ -0,0 +1,323 @@
+            return 1;
+        }
+        return 0;
+    }
+    int on_end_producer(Mlt::Producer*) { return 0; }
+    int on_start_playlist(Mlt::Playlist*) { return 0; }
+    int on_end_playlist(Mlt::Playlist*) { return 0; }
+    int on_start_tractor(Mlt::Tractor*) { return 0; }
+    int on_end_tractor(Mlt::Tractor*) { return 0; }
+    int on_start_multitrack(Mlt::Multitrack*) { return 0; }
+    int on_end_multitrack(Mlt::Multitrack*) { return 0; }
+    int on_start_track() { return 0; }
+    int on_end_track() { return 0; }
+    int on_end_filter(Mlt::Filter*) { return 0; }
+    int on_start_transition(Mlt::Transition*) { return 0; }
+    int on_end_transition(Mlt::Transition*) { return 0; }
+    int on_start_chain(Mlt::Chain*) { return 0; }
+    int on_end_chain(Mlt::Chain*) { return 0; }
+    int on_start_link(Mlt::Link*) { return 0; }
+    int on_end_link(Mlt::Link*) { return 0; }
+};
+
+static Mlt::Producer findProducer(const QUuid& uuid)
+{
+    FindProducerParser graphParser(uuid);
+    if (MAIN.isMultitrackValid()) {
+        graphParser.start(*MAIN.multitrack());
+        if (graphParser.producer().is_valid()) {
+            return graphParser.producer();
+        }
+    }
+    if (MAIN.playlist() && MAIN.playlist()->count() > 0) {
+        graphParser.start(*MAIN.playlist());
+        if (graphParser.producer().is_valid()) {
+            return graphParser.producer();
+        }
+    }
+    Mlt::Producer producer(MLT.isClip()? MLT.producer() : MLT.savedProducer());
+    if (producer.is_valid()) {
+        graphParser.start(producer);
+        if (graphParser.producer().is_valid()) {
+            return graphParser.producer();
+        }
+    }
+    return Mlt::Producer();
+}
+
+namespace Filter {
+
+AddCommand::AddCommand(AttachedFiltersModel& model, QmlMetadata* metadata, QUndoCommand* parent)
+    : QUndoCommand(parent)
+    , m_model(model)
+    , m_metadata(metadata)
+    , m_index(-1)
+    , m_row(-1)
+    , m_producer(*model.producer())
+    , m_producerUuid(MLT.ensureHasUuid(m_producer))
+{
+    Q_ASSERT(m_metadata);
+    setText(QObject::tr("Add %1 filter").arg(metadata->name()));
+}
+
+void AddCommand::redo()
+{
+    LOG_DEBUG() << text() << m_row;
+    if (m_producer.is_valid()) {
+        m_row = m_model.doAddService(m_metadata, m_producer, m_index);
+        // Only hold the producer reference for the first redo and lookup by UUID thereafter.
+        m_producer = Mlt::Producer();
+    } else {
+        Mlt::Producer producer = m_producer;
+        if (!producer.is_valid()) {
+            producer = findProducer(m_producerUuid);
+        }
+        Q_ASSERT(producer.is_valid());
+        Q_ASSERT(m_service.is_valid());
+        if (producer.is_valid()) {
+            m_model.doRestoreService(producer, m_service, m_index);
+        }
+    }
+}
+
+void AddCommand::undo()
+{
+    LOG_DEBUG() << text() << m_row;
+    Mlt::Producer producer(findProducer(m_producerUuid));
+    Q_ASSERT(producer.is_valid());
```

Comment:
In my testing, this assert triggers 100% of the time. I tried source clips, clips in timeline and tracks. All crash. If it does not trigger for you, I can try to debug it on my system.

Reply:
This is fixed.

Reply:
That fixed it for clips, But I still get the assert if I undo adding a filter to track or master (output)

Reply:
This is fixed for tracks and output, but I found another problem that also affects clips. Add a filter to a clip, remove the clip, undo that, and then undo the filter. The filter command is unable to find the producer because it was recreated when undoing the removal thereby losing the UUID the filter command had placed on it. Thus, I need to make a change so that all producers get tagged early on with a UUID that will serialize to XML when saved for QUndoCommands but removed when saved as a project. That will be a big change, and it may take a while to identify this everywhere including obscure spots or tricky follow-on bugs like ensuring a copy makes a new one, etc. So, now I am pessimistic about including this. I will likely give a couple of days and see how it goes.

Reply:
Instead of what I wrote above, it looks like I can change the UndoHelper and timeline and playlist QUndoCommands to save the UUID and reapply it to the producers they restore.

Reply:
OK, this latest problem (combination of filter and timeline commands) should be fixed now.